### PR TITLE
Update for version selection articles

### LINKED ISSUE #12237: Major-version roll-forward options for Core 3.0
Major-verison roll-forward has been added as an opt-in feature in Core 3.0. This should be documented in the version roll-foward article: dotnet/core/versions/selection.md

Here is info from the design spec at https://github.com/dotnet/designs/blob/main/accepted/2019/runtime-binding.md

### RollForward

`RollForward` specifies the roll-forward policy for an application, either as a fallback to accommodate missing a specific runtime version or as a directive to use a later version.

`RollForward` can have the following values:

* `LatestPatch` -- Roll forward to the highest patch version. This disables minor version roll forward.
* `Minor` -- Roll forward to the lowest higher minor version, if requested minor version is missing. If the requested minor version is present, then the `LatestPatch` policy is used.
* `Major` -- Roll forward to lowest higher major version, and lowest minor version, if requested major version is missing. If the requested major version is present, then the `Minor` policy is used.
* `LatestMinor` -- Roll forward to highest minor version, even if requested minor version is present.
* `LatestMajor` -- Roll forward to highest major and highest minor version, even if requested major is present.
* `Disable` -- Do not roll forward. Only bind to specified version. This policy is not recommended for general use since it disable the ability to roll-forward to the latest patches. It is only recommended for testing.

`Minor` is the default setting. See **Configuration Precedence** for more information.

In all cases except `Disable` the highest available patch version is selected.

Note: `LatestMinor` and `LatestMajor` are intended for component hosting scenarios, for both managed and native hosts (for example, managed COM components).

`RollForward` can be set in the following ways:

* Project file property: `RollForward`
* Runtime configuration file property: `rollForward`
* Environment variable: `DOTNET_ROLL_FORWARD`
* Command line argument: `--roll-forward`

edit by @mairaw: added topic details

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: ab345482-5752-6845-feee-52f0f06c620c
* Version Independent ID: 3613d4c0-be8e-c5a4-42f4-bec78ab99c75
* Content: [Select which .NET version to use - .NET](https://docs.microsoft.com/en-us/dotnet/core/versions/selection)
* Content Source: [docs/core/versions/selection.md](https://github.com/dotnet/docs/blob/master/docs/core/versions/selection.md)
* Product: **dotnet-core**
* GitHub Login: @Thraka
* Microsoft Alias: **adegeo**

### PR DESCRIPTION
**Pull request by adegeo**
## Summary

- Added MSBuild props for **RollFoward** and **RuntimeFrameworkVersion** ([internal preview](https://review.docs.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props?branch=pr-en-us-25039#rollforward)) @gewarren 
- Expanded **Select which .NET version to use** with more information about roll forward. Also updated examples to be .NET 5 focused instead of .NET Core 3.1 ([internal preview](https://review.docs.microsoft.com/en-us/dotnet/core/versions/selection?branch=pr-en-us-25039))
- Clarified **How .NET Runtime and SDK are versioned** ([internal preview](https://review.docs.microsoft.com/en-us/dotnet/core/versions/?branch=pr-en-us-25039))

Fixes #12237 

@vitek-karas @agocke

### CODE REVIEW

File: docs/core/project-sdk/msbuild-props.md [hunk 1]
```diff
@@ -330,6 +332,41 @@ The `PreserveCompilationReferences` property is similar to the [PreserveCompilat
...
+
+The `RuntimeFrameworkVersion` property specifies the version of the runtime to use when publishing. Specify a runtime version:
+
+``` xml
+<PropertyGroup>
+  <RuntimeFrameworkVersion>5.0.7</RuntimeFrameworkVersion>
+</PropertyGroup>
+```
+
+When publishing a framework-dependent application, this value specifies the minimum version required. When publishing a self-contained application, this value specifies the *exact* version required.
```
Comment (gewarren) [inline docs/core/project-sdk/msbuild-props.md:368]:
```suggestion
When publishing a framework-dependent application, this value specifies the *minimum* version required. When publishing a self-contained application, this value specifies the *exact* version required.
```

File: docs/core/project-sdk/msbuild-props.md [hunk 2]
```diff
@@ -330,6 +332,41 @@ The `PreserveCompilationReferences` property is similar to the [PreserveCompilat
...
+
+| Value         | Description                                                                                                                                                               |
+|---------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `LatestPatch` | Roll-forward to the highest patch version. This value disables minor version roll-forward.                                                                                      |
+| `Minor`       | Roll-forward to the lowest higher minor version, if requested minor version is missing. If the requested minor version is present, then the `LatestPatch` policy is used. |
+| `Major`       | Roll-forward to the next available higher major version, and lowest minor version, if requested major version is missing. If the requested major version is present, then the `Minor` policy is used. |
+| `LatestMinor` | Roll-forward to highest minor version, even if requested minor version is present.                                                                                        |
```
Comment (gewarren) [inline docs/core/project-sdk/msbuild-props.md:352]: Should we colocate all the `Latest*` values?

File: docs/core/versions/selection.md [hunk 3]
```diff
@@ -29,83 +29,151 @@ SDK commands include `dotnet new` and `dotnet run`. The .NET CLI must choose an
...
 
-1. `dotnet` searches for a *global.json* file iteratively reverse-navigating the path upward from the current working directory.
-1. `dotnet` uses the SDK specified in the first *global.json* found.
-1. `dotnet` uses the latest installed SDK if no *global.json* is found.
+01. `dotnet` searches for a *global.json* file iteratively reverse-navigating the path upward from the current working directory.
+01. `dotnet` uses the SDK specified in the first *global.json* found.
+01. `dotnet` uses the latest installed SDK if no *global.json* is found.
 
-You can learn more about selecting an SDK version in the [Matching rules](../tools/global-json.md#matching-rules) section of the article on *global.json*.
+For more information about SDK version selection, seethe [Matching rules](../tools/global-json.md#matching-rules) and [rollForward](../tools/global-json.md#rollforward) sections of the [global.json overview](../tools/global-json.md) article.
```
Comment (gewarren) [inline docs/core/versions/selection.md:54]:
```suggestion
For more information about SDK version selection, see the [Matching rules](../tools/global-json.md#matching-rules) and [rollForward](../tools/global-json.md#rollforward) sections of the [global.json overview](../tools/global-json.md) article.
```

File: docs/core/versions/selection.md [hunk 4]
```diff
@@ -29,83 +29,151 @@ SDK commands include `dotnet new` and `dotnet run`. The .NET CLI must choose an
...
-3. Later, the user installs 3.0.3 and runs the application again, 3.0.3 will now be used.
+01. The application specifies that 5.0 is required.
+02. When run, version 5.0.* isn't installed, however, 5.1.0 is. Version 3.1.0 will be used.
+03. Later, the user installs 5.0.3 and runs the application again, 5.0.3 will now be used.
+
+It's possible that 5.0.3 and 5.1.0 behave differently, particularly for scenarios like serializing binary data.
+
+### Control roll-forward behavior
+
+The roll-forward behavior for an application, can be configured in four different ways:
```
Comment (gewarren) [inline docs/core/versions/selection.md:103]:
```suggestion
The roll-forward behavior for an application can be configured in four different ways:
```

File: docs/core/versions/selection.md [hunk 5]
```diff
@@ -29,83 +29,151 @@ SDK commands include `dotnet new` and `dotnet run`. The .NET CLI must choose an
...
+
+    When you run an application, you can control the roll-forward behavior through the command line:
+
+    ```dotnet
+    dotnet run --roll-forward LatestMinor
+    dotnet myapp.dll --roll-forward LatestMinor
+    myapp.exe --roll-forward LatestMinor
+    ```
+
+04. The `DOTNET_ROLL_FORWARD` environment setting.
```
Comment (gewarren) [inline docs/core/versions/selection.md:140]:
```suggestion
04. The `DOTNET_ROLL_FORWARD` environment variable.
```

File: docs/core/versions/selection.md [hunk 6]
```diff
@@ -29,83 +29,151 @@ SDK commands include `dotnet new` and `dotnet run`. The .NET CLI must choose an
...
+
+### Precedence
+
+Roll forward behavior is considered in the following order:
+
+01. `*.runtimeconfig.json` config file.
+01. The `DOTNET_ROLL_FORWARD` environment setting.
```
Comment (gewarren) [inline docs/core/versions/selection.md:147]:
```suggestion
01. The `DOTNET_ROLL_FORWARD` environment variable.
```

File: docs/core/versions/selection.md [hunk 7]
```diff
@@ -29,83 +29,151 @@ SDK commands include `dotnet new` and `dotnet run`. The .NET CLI must choose an
...
+01. Any `--roll-forward` parameter passed to the running application.
+
+### Values
+
+However you set the roll-forward setting, use one of the following values to set the behavior:
 
-It's possible that 3.0.3 and 3.1.0 behave differently, particularly for scenarios like serializing binary data.
+| Value         | Description                                                                                                                                                               |
+|---------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `LatestPatch` | Roll-forward to the highest patch version. This value disables minor version roll-forward.                                                                                      |
```
Comment (gewarren) [inline docs/core/versions/selection.md:156]: Should we make this table an include file so we don't have to duplicate it?

File: docs/core/project-sdk/msbuild-props.md [hunk 8]
```diff
@@ -330,6 +332,41 @@ The `PreserveCompilationReferences` property is similar to the [PreserveCompilat
 
 For more information, see [Razor SDK properties](/aspnet/core/razor-pages/sdk#properties).
 
+### RollForward
+
+The `RollForward` property controls the how the application chooses a different runtime when the targetted runtime is unavailable. This value is output to the *.runtimeconfig.json* as the `rollForward` setting.
```
Comment (vitek-karas) [inline docs/core/project-sdk/msbuild-props.md:337]:
```suggestion
The `RollForward` property controls how the application chooses a runtime when multiple runtime versions are available. This value is output to the *.runtimeconfig.json* as the `rollForward` setting.
```

File: docs/core/project-sdk/msbuild-props.md [hunk 9]
```diff
@@ -330,6 +332,41 @@ The `PreserveCompilationReferences` property is similar to the [PreserveCompilat
...
+  <RollForward>LatestMinor</RollForward>
+</PropertyGroup>
+```
+
+Set `RollForward` to one of the following values:
+
+| Value         | Description                                                                                                                                                               |
+|---------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `LatestPatch` | Roll-forward to the highest patch version. This value disables minor version roll-forward.                                                                                      |
+| `Minor`       | Roll-forward to the lowest higher minor version, if requested minor version is missing. If the requested minor version is present, then the `LatestPatch` policy is used. |
```
Comment (vitek-karas) [inline docs/core/project-sdk/msbuild-props.md:350]: We should note that this is the default.

File: docs/core/versions/selection.md [hunk 10]
```diff
@@ -29,83 +29,151 @@ SDK commands include `dotnet new` and `dotnet run`. The .NET CLI must choose an
...
+    dotnet run --roll-forward LatestMinor
+    dotnet myapp.dll --roll-forward LatestMinor
+    myapp.exe --roll-forward LatestMinor
+    ```
+
+04. The `DOTNET_ROLL_FORWARD` environment setting.
+
+### Precedence
+
+Roll forward behavior is considered in the following order:
```
Comment (vitek-karas) [inline docs/core/versions/selection.md:144]: This is confusing (at least to me). Reading this I would assume the `runtimeconfig.json` is the higher priority and thus overrides all of the other options, but it's actually the other way round. I think there should be an explanation of what the order in this list means.

**Comment by vitek-karas** [comment]
Thanks a lot for making these changes!

File: docs/core/versions/selection.md (hunk 10 above)
Comment (adegeo) [reply docs/core/versions/selection.md:144]: OK, I added some descriptions on each of the items.

**Review by gewarren — APPROVED** [review]

**Pull request merged**